Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp

```diff
@@ -393,17 +395,74 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<JS::ArrayBuffer>> RSAOAEP::encrypt(Algorith
+                auto name_property = TRY(obj->get("name"));
+                return name_property.to_string(realm.vm()); }));
+
+    auto const& handle = key->handle();
+
+    auto maybe_data = handle.visit(
+        [&](::Crypto::PK::RSAPublicKey<> const& public_key) -> ErrorOr<ByteBuffer> {
+            auto error_message = MUST(String::formatted("Invalid hash function '{}'", hash));
+            ErrorOr<ByteBuffer> maybe_padding = Error::from_string_view(error_message.bytes_as_string_view());
+
+            if (hash.equals_ignoring_ascii_case("SHA-1"sv)) {
+                maybe_padding = ::Crypto::Padding::OAEP::eme_encode<::Crypto::Hash::SHA1, ::Crypto::Hash::MGF>(
+                    plaintext,
+                    label,
+                    public_key.length());
+            } else if (hash.equals_ignoring_ascii_case("SHA-256"sv)) {
+                maybe_padding = ::Crypto::Padding::OAEP::eme_encode<::Crypto::Hash::SHA256, ::Crypto::Hash::MGF>(
+                    plaintext,
+                    label,
+                    public_key.length());
+            } else if (hash.equals_ignoring_ascii_case("SHA-384"sv)) {
+                maybe_padding = ::Crypto::Padding::OAEP::eme_encode<::Crypto::Hash::SHA384, ::Crypto::Hash::MGF>(
+                    plaintext,
+                    label,
+                    public_key.length());
+            } else if (hash.equals_ignoring_ascii_case("SHA-512"sv)) {
+                maybe_padding = ::Crypto::Padding::OAEP::eme_encode<::Crypto::Hash::SHA512, ::Crypto::Hash::MGF>(
+                    plaintext,
+                    label,
+                    public_key.length());
+            }
+
+            if (maybe_padding.is_error()) {
+                dbgln("RSAOAEP::encrypt: Error encoding padding: {}", maybe_padding.error());
+                return maybe_padding;
+            }
+
+            auto padding = maybe_padding.release_value();
+
+            auto output_buffer = TRY(ByteBuffer::create_uninitialized(public_key.length()));
+            auto output_bytes = output_buffer.bytes();
+
+            auto rsa = ::Crypto::PK::RSA {};
+            rsa.set_public_key(public_key);
+            rsa.encrypt(padding, output_bytes);
+
+            return output_buffer;
+        },
+        [](auto) -> ErrorOr<ByteBuffer> {
+            VERIFY_NOT_REACHED();
+        });
 
     // 4. If performing the operation results in an error, then throw an OperationError.
+    if (maybe_data.is_error()) {
+        auto error_message = MUST(FlyString::from_utf8(maybe_data.error().string_literal()));
+        return WebIDL::OperationError::create(realm, error_message);
+    }
+
+    auto data = maybe_data.release_value();
 
     // 5. Let ciphertext be the value C that results from performing the operation.
-    // FIXME: Actually encrypt the data
-    auto ciphertext = TRY_OR_THROW_OOM(vm, ByteBuffer::copy(plaintext));
+    auto ciphertext = TRY_OR_THROW_OOM(vm, ByteBuffer::copy(data));
```

Comment:
we don't actually need to copy this here, that was just there to put some stub code in before the encryption worked.

```
auto ciphertext = maybe_data.release_value();
```